.github/scripts/utils.py/update-crowdin-locales.py

```diff
@@ -0,0 +1,88 @@
+import os
+import logging
+import json
+
+from crowdin_api import CrowdinClient
+from fnmatch import fnmatch
+
+CROWDIN_PROJECT_ID = 51028
+CROWDIN_API_TOKEN = os.environ["CROWDIN_PAT"]
+DEFAULT_LOCALE = "en-US"
+
+crowdin_api = CrowdinClient(token=CROWDIN_API_TOKEN, project_id=CROWDIN_PROJECT_ID, page_size=500)
+
+logging.basicConfig()
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+def get_dir_id(name: str) -> int:
+    crowdin_api.source_files.list_directories(filter=name)["data"][0]["data"]["id"]
+
+
+def add_to_crowdin_storage(file_path: str) -> int:
+    return crowdin_api.storages.add_storage(open(file_path))["data"]["id"]
+
+
+def update_crowdin_file(file_id: int, file_path: str) -> None:
+    storage_id = add_to_crowdin_storage(file_path)
+    crowdin_api.source_files.update_file(file_id, storage_id)
+
+    logger.info(f"{file_path} updated on Crowdin.")
+
+
+def create_crowdin_file(file_path: str, name: str, directory_name: str, export_pattern: str):
+    storage_id = add_to_crowdin_storage(file_path)
+
+    crowdin_api.source_files.add_file(storage_id,
+                                      name,
+                                      directoryId=get_dir_id(directory_name),
+                                      exportOptions={"exportPattern": export_pattern})
+
+    logger.info(f"{file_path} created on Crowdin in {directory_name} as {name}.")
+
+
+def upload(updated_locales: list[str]) -> None:
+    export_paths_map = dict()
+
+    for source_file_data in crowdin_api.source_files.list_files()["data"]:
+        source_file = source_file_data["data"]
+
+        if "exportOptions" not in source_file:
+            continue
+
+        export_path: str = source_file["exportOptions"]["exportPattern"]
```

Comment:
Multi-dimensional direct access will lead to `ValueError` if key doesn't return dictionary with the expected key. Use chained `get` calls to ensure the access does not fail and use a fallback like `None` to check for errors.

Reply:
If the dictionary access fails, there is no reason to recover from there on. If there is no `exportPattern` in `exportOptions`, then the configuration on Crowdin needs to be updated. The code depends a lot on the configuration on Crowdin and assumes a specific structure. I support making the code flexible to the API results, but have trouble understanding what makes encapsulating all possible situations with error handling because of possible misconfiguration on Crowdin or broken responses beneficial.
Adding optional debug logging for all responses makes in my head more sense.

Reply:
Every time you access an API or process data you're not in control of, you have to ensure to either safely evaluate the state of the data before accessing it to stay in control of program execution (otherwise you openly rely on undefined state and you can make no assumptions of the states your code can end up in).

If _any_ issue while accessing that data is considered a possible error, at the very least wrap the access in a `try: [...] except (ValueError, KeyError) as e:` block because the code can potentially break on both of these errors because of an invalid access violation. You can then handle the error by logging an appropriate error function, clean up, and exit the program with an error code under your control.